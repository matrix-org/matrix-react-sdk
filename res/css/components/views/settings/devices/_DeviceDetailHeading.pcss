/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.mx_DeviceDetailHeading {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
}

.mx_DeviceDetailHeading_renameCta {
    flex-shrink: 0;
}

.mx_DeviceDetailHeading_renameForm {
    display: grid;
    grid-gap: $spacing-16;
    justify-content: left;
    grid-template-columns: 100%;
}

.mx_DeviceDetailHeading_renameFormButtons {
    display: flex;
    flex-direction: row;
    gap: $spacing-8;

    .mx_Spinner {
        width: auto;
        flex-grow: 0;
    }
}

.mx_DeviceDetailHeading_renameFormInput {
    // override field styles
    margin: 0 0 $spacing-4 0 !important;
}

.mx_DeviceDetailHeading_renameFormHeading {
    margin: 0;
}

.mx_DeviceDetailHeading_renameFormError {
    color: $alert;
    padding-right: $spacing-4;
    display: block;
}
