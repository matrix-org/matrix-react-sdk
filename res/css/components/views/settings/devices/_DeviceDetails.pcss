/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.mx_DeviceDetails {
    display: flex;
    flex-direction: column;
    box-sizing: border-box;

    width: 100%;

    margin-top: $spacing-16;
    padding: $spacing-16;
    border-radius: 8px;
    border: 1px solid $quinary-content;
}

.mx_DeviceDetails_section {
    padding-bottom: $spacing-16;
    margin-bottom: $spacing-16;
    border-bottom: 1px solid $quinary-content;

    display: grid;
    grid-gap: $spacing-16;

    &:last-child {
        padding-bottom: 0;
        border-bottom: 0;
        margin-bottom: 0;
    }
}

.mx_DeviceDetails_sectionHeading {
    margin: 0;
}

.mxDeviceDetails_metadataTable {
    font-size: $font-12px;
    color: $secondary-content;

    width: 100%;

    border-spacing: 0;

    th {
        text-transform: uppercase;
        font-weight: normal;
        text-align: left;
    }

    td {
        padding-top: $spacing-8;
    }

    .mxDeviceDetails_metadataLabel {
        width: 160px;
    }

    .mxDeviceDetails_metadataValue {
        color: $primary-content;
    }
}
