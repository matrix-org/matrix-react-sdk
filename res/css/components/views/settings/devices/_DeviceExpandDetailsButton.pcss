/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.mx_DeviceExpandDetailsButton {
    height: 32px;
    width: 32px;
    background: transparent;

    border-radius: 4px;
    color: $icon-button-color;

    --icon-transform: rotate(-90deg);

    &:hover {
        background: $quinary-content;
    }
}

.mx_DeviceExpandDetailsButton.mx_DeviceExpandDetailsButton_expanded {
    --icon-transform: rotate(0deg);

    background: $system;

    &:hover {
        background: $quinary-content;
    }
}

.mx_DeviceExpandDetailsButton_icon {
    height: 16px;
    width: 16px;

    transition: all 0.3s;
    transform: var(--icon-transform);
    transform-origin: center;
}
