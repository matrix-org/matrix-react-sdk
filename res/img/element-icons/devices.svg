<svg width="32" height="32" viewBox="0 0 32 32" fill="none" xmlns="http://www.w3.org/2000/svg">
  <g clip-path="url(#clip0_721_35674)">
  <path d="M5.33333 9.33313C5.33333 8.5998 5.93333 7.9998 6.66667 7.9998H28C28.7333 7.9998 29.3333 7.3998 29.3333 6.66646C29.3333 5.93313 28.7333 5.33313 28 5.33313H5.33333C3.86667 5.33313 2.66667 6.53313 2.66667 7.9998V22.6665H2C0.893333 22.6665 0 23.5598 0 24.6665C0 25.7731 0.893333 26.6665 2 26.6665H18.6667V22.6665H5.33333V9.33313ZM30.6667 10.6665H22.6667C21.9333 10.6665 21.3333 11.2665 21.3333 11.9998V25.3331C21.3333 26.0665 21.9333 26.6665 22.6667 26.6665H30.6667C31.4 26.6665 32 26.0665 32 25.3331V11.9998C32 11.2665 31.4 10.6665 30.6667 10.6665ZM29.3333 22.6665H24V13.3331H29.3333V22.6665Z" fill="currentColor"/>
  </g>
  <defs>
  <clipPath id="clip0_721_35674">
  <rect width="32" height="32" fill="white"/>
  </clipPath>
  </defs>
  </svg>
  