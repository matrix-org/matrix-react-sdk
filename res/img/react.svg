<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
