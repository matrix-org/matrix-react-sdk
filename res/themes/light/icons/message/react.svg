<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>