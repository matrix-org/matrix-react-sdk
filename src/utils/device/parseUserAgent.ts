/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import UAParser from "ua-parser-js";

export enum DeviceType {
    Desktop = "Desktop",
    Mobile = "Mobile",
    Web = "Web",
    Unknown = "Unknown",
}
export type ExtendedDeviceInformation = {
    deviceType: DeviceType;
    // eg Google Pixel 6
    deviceModel?: string;
    // eg Android 11
    deviceOperatingSystem?: string;
    // eg Firefox 1.1.0
    client?: string;
};

// Element/1.8.21 (iPhone XS Max; iOS 15.2; Scale/3.00)
const IOS_KEYWORD = "; iOS ";
const BROWSER_KEYWORD = "Mozilla/";

const getDeviceType = (
    userAgent: string,
    device: UAParser.IDevice,
    browser: UAParser.IBrowser,
    operatingSystem: UAParser.IOS,
): DeviceType => {
    if (browser.name === "Electron") {
        return DeviceType.Desktop;
    }
    if (!!browser.name) {
        return DeviceType.Web;
    }
    if (device.type === "mobile" || operatingSystem.name?.includes("Android") || userAgent.indexOf(IOS_KEYWORD) > -1) {
        return DeviceType.Mobile;
    }
    return DeviceType.Unknown;
};

interface CustomValues {
    customDeviceModel?: string;
    customDeviceOS?: string;
}
/**
 * Some mobile model and OS strings are not recognised
 * by the UA parsing library
 * check they exist by hand
 */
const checkForCustomValues = (userAgent: string): CustomValues => {
    if (userAgent.includes(BROWSER_KEYWORD)) {
        return {};
    }

    const mightHaveDevice = userAgent.includes("(");
    if (!mightHaveDevice) {
        return {};
    }
    const deviceInfoSegments = userAgent.substring(userAgent.indexOf("(") + 1).split("; ");
    const customDeviceModel = deviceInfoSegments[0] || undefined;
    const customDeviceOS = deviceInfoSegments[1] || undefined;
    return { customDeviceModel, customDeviceOS };
};

const concatenateNameAndVersion = (name?: string, version?: string): string | undefined =>
    name && [name, version].filter(Boolean).join(" ");

export const parseUserAgent = (userAgent?: string): ExtendedDeviceInformation => {
    if (!userAgent) {
        return {
            deviceType: DeviceType.Unknown,
        };
    }

    const parser = new UAParser(userAgent);

    const browser = parser.getBrowser();
    const device = parser.getDevice();
    const operatingSystem = parser.getOS();

    const deviceType = getDeviceType(userAgent, device, browser, operatingSystem);

    // OSX versions are frozen at 10.15.17 in UA strings https://chromestatus.com/feature/5452592194781184
    // ignore OS version in browser based sessions
    const shouldIgnoreOSVersion = deviceType === DeviceType.Web || deviceType === DeviceType.Desktop;
    const deviceOperatingSystem = concatenateNameAndVersion(
        operatingSystem.name,
        shouldIgnoreOSVersion ? undefined : operatingSystem.version,
    );
    const deviceModel = concatenateNameAndVersion(device.vendor, device.model);
    const client = concatenateNameAndVersion(browser.name, browser.version);

    // only try to parse custom model and OS when device type is known
    const { customDeviceModel, customDeviceOS } =
        deviceType !== DeviceType.Unknown ? checkForCustomValues(userAgent) : ({} as CustomValues);

    return {
        deviceType,
        deviceModel: deviceModel || customDeviceModel,
        deviceOperatingSystem: deviceOperatingSystem || customDeviceOS,
        client,
    };
};
