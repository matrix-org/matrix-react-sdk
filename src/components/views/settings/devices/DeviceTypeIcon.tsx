/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import classNames from "classnames";

import { Icon as UnknownDeviceIcon } from "../../../../../res/img/element-icons/settings/unknown-device.svg";
import { Icon as DesktopIcon } from "../../../../../res/img/element-icons/settings/desktop.svg";
import { Icon as WebIcon } from "../../../../../res/img/element-icons/settings/web.svg";
import { Icon as MobileIcon } from "../../../../../res/img/element-icons/settings/mobile.svg";
import { Icon as VerifiedIcon } from "../../../../../res/img/e2e/verified.svg";
import { Icon as UnverifiedIcon } from "../../../../../res/img/e2e/warning.svg";
import { _t, _td, TranslationKey } from "../../../../languageHandler";
import { ExtendedDevice } from "./types";
import { DeviceType } from "../../../../utils/device/parseUserAgent";

interface Props {
    isVerified?: ExtendedDevice["isVerified"];
    isSelected?: boolean;
    deviceType?: DeviceType;
}

const deviceTypeIcon: Record<DeviceType, React.FC<React.SVGProps<SVGSVGElement>>> = {
    [DeviceType.Desktop]: DesktopIcon,
    [DeviceType.Mobile]: MobileIcon,
    [DeviceType.Web]: WebIcon,
    [DeviceType.Unknown]: UnknownDeviceIcon,
};
const deviceTypeLabel: Record<DeviceType, TranslationKey> = {
    [DeviceType.Desktop]: _td("settings|sessions|desktop_session"),
    [DeviceType.Mobile]: _td("settings|sessions|mobile_session"),
    [DeviceType.Web]: _td("settings|sessions|web_session"),
    [DeviceType.Unknown]: _td("settings|sessions|unknown_session"),
};

export const DeviceTypeIcon: React.FC<Props> = ({ isVerified, isSelected, deviceType }) => {
    const Icon = deviceTypeIcon[deviceType!] || deviceTypeIcon[DeviceType.Unknown];
    const label = _t(deviceTypeLabel[deviceType!] || deviceTypeLabel[DeviceType.Unknown]);
    return (
        <div
            className={classNames("mx_DeviceTypeIcon", {
                mx_DeviceTypeIcon_selected: isSelected,
            })}
        >
            <div className="mx_DeviceTypeIcon_deviceIconWrapper">
                <Icon className="mx_DeviceTypeIcon_deviceIcon" role="img" aria-label={label} />
            </div>
            {isVerified ? (
                <VerifiedIcon
                    className={classNames("mx_DeviceTypeIcon_verificationIcon", "verified")}
                    role="img"
                    aria-label={_t("common|verified")}
                />
            ) : (
                <UnverifiedIcon
                    className={classNames("mx_DeviceTypeIcon_verificationIcon", "unverified")}
                    role="img"
                    aria-label={_t("common|unverified")}
                />
            )}
        </div>
    );
};
