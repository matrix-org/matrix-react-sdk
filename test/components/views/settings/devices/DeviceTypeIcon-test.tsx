/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { render } from "@testing-library/react";
import React from "react";

import { DeviceTypeIcon } from "../../../../../src/components/views/settings/devices/DeviceTypeIcon";
import { DeviceType } from "../../../../../src/utils/device/parseUserAgent";

describe("<DeviceTypeIcon />", () => {
    const defaultProps = {
        isVerified: false,
        isSelected: false,
    };
    const getComponent = (props = {}) => <DeviceTypeIcon {...defaultProps} {...props} />;

    it("renders an unverified device", () => {
        const { container } = render(getComponent());
        expect(container).toMatchSnapshot();
    });

    it("renders a verified device", () => {
        const { container } = render(getComponent({ isVerified: true }));
        expect(container).toMatchSnapshot();
    });

    it("renders correctly when selected", () => {
        const { container } = render(getComponent({ isSelected: true }));
        expect(container).toMatchSnapshot();
    });

    it("renders an unknown device icon when no device type given", () => {
        const { getByLabelText } = render(getComponent());
        expect(getByLabelText("Unknown session type")).toBeTruthy();
    });

    it("renders a desktop device type", () => {
        const deviceType = DeviceType.Desktop;
        const { getByLabelText } = render(getComponent({ deviceType }));
        expect(getByLabelText("Desktop session")).toBeTruthy();
    });

    it("renders a web device type", () => {
        const deviceType = DeviceType.Web;
        const { getByLabelText } = render(getComponent({ deviceType }));
        expect(getByLabelText("Web session")).toBeTruthy();
    });

    it("renders a mobile device type", () => {
        const deviceType = DeviceType.Mobile;
        const { getByLabelText } = render(getComponent({ deviceType }));
        expect(getByLabelText("Mobile session")).toBeTruthy();
    });

    it("renders an unknown device type", () => {
        const deviceType = DeviceType.Unknown;
        const { getByLabelText } = render(getComponent({ deviceType }));
        expect(getByLabelText("Unknown session type")).toBeTruthy();
    });
});
