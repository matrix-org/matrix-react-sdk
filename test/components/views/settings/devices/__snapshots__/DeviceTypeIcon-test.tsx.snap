// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeviceTypeIcon /> renders a verified device 1`] = `
<div>
  <div
    class="mx_DeviceTypeIcon"
  >
    <div
      aria-label="Unknown device type"
      class="mx_DeviceTypeIcon_deviceIcon"
      role="img"
    />
    <div
      aria-label="Verified"
      class="mx_DeviceTypeIcon_verificationIcon verified"
      role="img"
    />
  </div>
</div>
`;

exports[`<DeviceTypeIcon /> renders an unverified device 1`] = `
<div>
  <div
    class="mx_DeviceTypeIcon"
  >
    <div
      aria-label="Unknown device type"
      class="mx_DeviceTypeIcon_deviceIcon"
      role="img"
    />
    <div
      aria-label="Unverified"
      class="mx_DeviceTypeIcon_verificationIcon unverified"
      role="img"
    />
  </div>
</div>
`;

exports[`<DeviceTypeIcon /> renders correctly when selected 1`] = `
<div>
  <div
    class="mx_DeviceTypeIcon mx_DeviceTypeIcon_selected"
  >
    <div
      aria-label="Unknown device type"
      class="mx_DeviceTypeIcon_deviceIcon"
      role="img"
    />
    <div
      aria-label="Unverified"
      class="mx_DeviceTypeIcon_verificationIcon unverified"
      role="img"
    />
  </div>
</div>
`;
